doc/codegen/modular_codegen_design.md

```diff
@@ -0,0 +1,561 @@
+
+## Organizational Structure
+
+### Current Structure
+```
+dace/codegen/
+├── __init__.py
+├── codegen.py              # Main entry point
+├── targets/                # Target-specific generators
+├── instrumentation/        # Profiling support
+├── tools/                  # Utilities
+├── dispatcher.py           # Target routing
+├── control_flow.py         # Control flow analysis
+├── compiler.py             # CMake integration
+├── codeobject.py          # Code file representation
+├── compiled_sdfg.py       # Runtime interface
+└── ...
+```
+
+### Proposed Structure
+```
+dace/codegen/
+├── __init__.py
+├── codegen.py              # Simplified entry point
+├── compiler/               # Compilation and build system
+│   ├── __init__.py
+│   ├── cmake.py            # CMake integration (from compiler.py)
+│   ├── direct.py           # Direct compiler calls
+│   ├── build_system.py     # Abstract build system interface
+│   └── environments.py     # Environment-specific build logic
+├── passes/                 # Code generation passes
+│   ├── __init__.py
+│   ├── analysis/           # Analysis passes
+│   │   ├── __init__.py
+│   │   ├── type_inference.py
+│   │   ├── metadata_collection.py
+│   │   ├── allocation_analysis.py
+│   │   ├── control_flow_analysis.py
+│   │   └── target_analysis.py
+│   ├── transformation/     # Transformation passes
+│   │   ├── __init__.py
+│   │   ├── copy_to_map.py
+│   │   ├── stream_assignment.py
+│   │   ├── tasklet_lowering.py
+│   │   └── memlet_lowering.py
+│   ├── codegen/           # Code generation passes
+│   │   ├── __init__.py
+│   │   ├── state_struct.py
+│   │   ├── allocation_code.py
+│   │   ├── frame_code.py
+│   │   ├── target_code.py
+│   │   └── header_generation.py
+│   ├── file_generation/   # File organization passes
+│   │   ├── __init__.py
+│   │   ├── sdfg_splitting.py
+│   │   └── code_objects.py
+│   └── pipelines.py       # Pre-built pipelines
+├── targets/               # Target-specific generators (simplified)
+│   ├── __init__.py
+│   ├── base.py            # Base target interface (from target.py)
+│   ├── openmp.py          # OpenMP backend (split from cpu.py)
+│   ├── cpp.py             # Pure C++ backend
+│   ├── gpu.py             # GPU backend (generalized from cuda.py)
+│   ├── cuda.py            # CUDA-specific GPU
+│   ├── hip.py             # HIP-specific GPU
+│   ├── fpga/              # FPGA backends
+│   └── specialized/       # Other specialized targets
+├── runtime/               # Runtime interface (from compiled_sdfg.py)
+└── utils/                 # Utilities (dispatcher, codeobject, etc.)
+    ├── __init__.py
+    ├── dispatcher.py
+    ├── codeobject.py
+    ├── prettycode.py
+    └── common.py
+```
+
+## Target Refactoring
+
+### Current Issues
+- **CPU backend** actually does OpenMP generation
+- **CUDA backend** is GPU-specific, not general GPU
+- Poor separation between generic and specific backends
+
+### Proposed Refactoring
```

Comment:
What is the purpose and API of targets in this new refactoring? Which passes use them? This should be documented in the design document.